redundant closure lint fails for FnMut where FnOnce is expected

This compiles ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7e067bffb234f1b39857d6dd986e1313)):

```rust
pub fn bar(mut x: Box<dyn FnMut()>) {
    foo(|| x())
}

pub fn foo<T: FnOnce()>(_: T) {}
```

but clippy suggests:

```
warning: redundant closure found
 --> src/lib.rs:2:9
  |
2 |     foo(|| x())
  |         ^^^^^^ help: remove closure as shown: `x`
  |
  = note: #[warn(clippy::redundant_closure)] on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

which fails to compile with ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bd910962f25b592909b97a16c5198686)):

```
error[E0277]: expected a `std::ops::FnOnce<()>` closure, found `std::boxed::Box<dyn std::ops::FnMut()>`
 --> src/lib.rs:2:5
  |
2 |     foo(x)
  |     ^^^ expected an `FnOnce<()>` closure, found `std::boxed::Box<dyn std::ops::FnMut()>`
  |
  = help: the trait `std::ops::FnOnce<()>` is not implemented for `std::boxed::Box<dyn std::ops::FnMut()>`
  = note: wrap the `std::boxed::Box<dyn std::ops::FnMut()>` in a closure with no arguments: `|| { /* code */ }
note: required by `foo`
 --> src/lib.rs:5:1
  |
5 | pub fn foo<T: FnOnce()>(_: T) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

--- a/clippy_lints/src/eta_reduction.rs
+++ b/clippy_lints/src/eta_reduction.rs
@@ -1,4 +1,5 @@
 use if_chain::if_chain;
+use matches::matches;
 use rustc::hir::*;
 use rustc::lint::{in_external_macro, LateContext, LateLintPass, LintArray, LintContext, LintPass};
 use rustc::ty::{self, Ty};
@@ -66,7 +67,7 @@ fn check_closure(cx: &LateContext<'_, '_>, expr: &Expr) {
 
             let fn_ty = cx.tables.expr_ty(caller);
 
-            if let ty::FnDef(_, _) = fn_ty.sty;
+            if matches!(fn_ty.sty, ty::FnDef(_, _) | ty::FnPtr(_) | ty::Closure(_, _));
 
             if !type_is_unsafe_function(cx, fn_ty);
 

--- a/clippy_lints/src/eta_reduction.rs
+++ b/clippy_lints/src/eta_reduction.rs
@@ -65,6 +65,9 @@ fn check_closure(cx: &LateContext<'_, '_>, expr: &Expr) {
             if !(is_adjusted(cx, ex) || args.iter().any(|arg| is_adjusted(cx, arg)));
 
             let fn_ty = cx.tables.expr_ty(caller);
+
+            if let ty::FnDef(_, _) = fn_ty.sty;
+
             if !type_is_unsafe_function(cx, fn_ty);
 
             if compare_inputs(&mut iter_input_pats(decl, body), &mut args.into_iter());

--- a/tests/ui/eta.fixed
+++ b/tests/ui/eta.fixed
@@ -138,3 +138,14 @@ fn passes_fn_mut(mut x: Box<dyn FnMut()>) {
     requires_fn_once(|| x());
 }
 fn requires_fn_once<T: FnOnce()>(_: T) {}
+
+fn test_redundant_closure_with_function_pointer() {
+    type FnPtrType = fn(u8);
+    let foo_ptr: FnPtrType = foo;
+    let a = Some(1u8).map(foo_ptr);
+}
+
+fn test_redundant_closure_with_another_closure() {
+    let closure = |a| println!("{}", a);
+    let a = Some(1u8).map(closure);
+}

--- a/tests/ui/eta.fixed
+++ b/tests/ui/eta.fixed
@@ -133,3 +133,8 @@ fn divergent(_: u8) -> ! {
 fn generic<T>(_: T) -> u8 {
     0
 }
+
+fn passes_fn_mut(mut x: Box<dyn FnMut()>) {
+    requires_fn_once(|| x());
+}
+fn requires_fn_once<T: FnOnce()>(_: T) {}

--- a/tests/ui/eta.rs
+++ b/tests/ui/eta.rs
@@ -138,3 +138,14 @@ fn passes_fn_mut(mut x: Box<dyn FnMut()>) {
     requires_fn_once(|| x());
 }
 fn requires_fn_once<T: FnOnce()>(_: T) {}
+
+fn test_redundant_closure_with_function_pointer() {
+    type FnPtrType = fn(u8);
+    let foo_ptr: FnPtrType = foo;
+    let a = Some(1u8).map(|a| foo_ptr(a));
+}
+
+fn test_redundant_closure_with_another_closure() {
+    let closure = |a| println!("{}", a);
+    let a = Some(1u8).map(|a| closure(a));
+}

--- a/tests/ui/eta.rs
+++ b/tests/ui/eta.rs
@@ -133,3 +133,8 @@ fn divergent(_: u8) -> ! {
 fn generic<T>(_: T) -> u8 {
     0
 }
+
+fn passes_fn_mut(mut x: Box<dyn FnMut()>) {
+    requires_fn_once(|| x());
+}
+fn requires_fn_once<T: FnOnce()>(_: T) {}

--- a/tests/ui/eta.stderr
+++ b/tests/ui/eta.stderr
@@ -68,5 +68,17 @@ error: redundant closure found
 LL |     let e: std::vec::Vec<char> = vec!['a', 'b', 'c'].iter().map(|c| c.to_ascii_uppercase()).collect();
    |                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove closure as shown: `char::to_ascii_uppercase`
 
-error: aborting due to 11 previous errors
+error: redundant closure found
+  --> $DIR/eta.rs:145:27
+   |
+LL |     let a = Some(1u8).map(|a| foo_ptr(a));
+   |                           ^^^^^^^^^^^^^^ help: remove closure as shown: `foo_ptr`
+
+error: redundant closure found
+  --> $DIR/eta.rs:150:27
+   |
+LL |     let a = Some(1u8).map(|a| closure(a));
+   |                           ^^^^^^^^^^^^^^ help: remove closure as shown: `closure`
+
+error: aborting due to 13 previous errors
 

